File: frameworks/unit-or-qa/junit-vanilla/src/test/java/io/nyrkio/benchzoo/junit/SampleTest.java
package io.nyrkio.benchzoo.junit;

import org.junit.jupiter.api.Test;

import java.io.FileOutputStream;
import java.io.IOException;
import java.time.ZoneOffset;
import java.time.ZonedDateTime;
import java.util.Random;

/**
 * Raw JUnit 5 implementation of the canonical benchzoo sample benchmark.
 *
 * <p>Four plain {@code @Test} methods — no JMH, no AssertJ benchmarking,
 * no pytest-benchmark-style timing extension. The JUnit test runner
 * (driven by Maven Surefire) records per-test wall-clock duration in its
 * junit XML output regardless, and that duration is what benchzoo's
 * {@code junit_vanilla} parser consumes.
 *
 * <p>See {@code docs/sample-benchmark.md} for the spec. Test method
 * names (benchmark1..benchmark4) are kept identical to the canonical
 * names so the parser can use them directly as
 * {@code attributes["test_name"]} with no normalization.
 */
public class SampleTest {

    /**
     * Seeded RNG for benchmark3, to keep the random payload deterministic
     * run-to-run. The actual byte values do not matter for wall-clock
     * measurement, but a fixed seed makes differential debugging easier.
     */
    private final Random random = new Random(42);

    /** Test 1 — sleep-dominated (~2.15 s). */
    @Test
    public void benchmark1() throws InterruptedException {
        Thread.sleep(2150);
    }

    /**
     * Test 2 — tight CPU loop, sub-millisecond.
     *
     * <p>Unlike the JMH implementation, there is no {@code Blackhole} to
     * defeat dead-code elimination here. JUnit tests are not run under
     * a benchmark harness; the JIT is free to prove the loop has no
     * observable side effects and delete it. We accept that: test 2's
     * role in the corpus is to exercise sub-millisecond measurement
     * handling in the parser, not to produce a physically meaningful
     * loop time. A volatile sink keeps the loop from being trivially
     * hoisted to a constant.
     */
    @Test
    public void benchmark2() {
        int sink = 0;
        for (int i = 0; i < 1000; i++) {
            sink += i;
        }
        // Use the value so the compiler can't prove it's dead.
        if (sink == Integer.MIN_VALUE) {
            throw new AssertionError("unreachable");
        }
    }

    /** Test 3 — write 1.4 MB of pseudo-random data to /dev/null. */
    @Test
    public void benchmark3() throws IOException {
        byte[] payload = new byte[1_400_000];
        random.nextBytes(payload);
        try (FileOutputStream out = new FileOutputStream("/dev/null")) {
            out.write(payload);
        }
    }

    /**
     * Test 4 — monthly change-point showcase. Sleeps for
     * {@code 2.15 + ((month mod 3) - 1)} seconds, where {@code month}
     * is the current UTC month. See {@code docs/sample-benchmark.md}
     * for the formula table.
     */
    @Test
    public void benchmark4() throws InterruptedException {
        int month = ZonedDateTime.now(ZoneOffset.UTC).getMonthValue();
        long sleepMillis = (long) ((2.15 + (month % 3 - 1)) * 1000);
        Thread.sleep(sleepMillis);
    }
}

File: frameworks/unit-or-qa/mocha/test/sample.test.js
// Canonical sample benchmark expressed as four Mocha unit tests.
//
// Mocha itself is a unit-test runner, not a benchmark framework. We use it
// here as a "unit test runner as timing source" per
// docs/parser-targets.md section 6: mocha's native `--reporter json` emits
// one JSON document with per-test `duration` (in milliseconds) for every
// test, and when test names are stable that is a perfectly usable
// performance time series.
//
// NOTE: this file uses mocha's native JSON reporter — NOT mocha-junit-reporter.
// The captured output is a single JSON document, not JUnit XML. See the
// README's "Parser notes" section for the shape.

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

describe('sample benchmark', function () {
  // Per-test timeout bumped above mocha's 2 s default so the sleep-dominated
  // tests (1 and 4, up to 3.15 s) don't hit the timeout.
  this.timeout(10000);

  it('benchmark1', async function () {
    // Sleep-dominated: ~2.15 s wall time.
    await sleep(2150);
  });

  it('benchmark2', function () {
    // Tight CPU loop, sub-millisecond. JavaScript engines can optimize
    // empty loops; touching `sink` keeps the body observable.
    let sink = 0;
    for (let i = 0; i < 1000; i++) {
      sink += i;
    }
    if (sink !== 499500) throw new Error('sink mismatch');
  });

  it('benchmark3', function () {
    // Allocate and fill 1.4 MB. Node has no direct /dev/null write API
    // worth using here; allocating and filling an ArrayBuffer is the
    // closest in-process analogue and exercises the same I/O-shaped
    // timing bucket as the bash reference.
    const N = 1_400_000;
    const buf = new Uint8Array(N);
    for (let i = 0; i < N; i++) {
      buf[i] = i & 0xff;
    }
    if (buf.length !== N) throw new Error('length mismatch');
  });

  it('benchmark4', async function () {
    // Monthly change-point showcase. sleep_s = 2.15 + ((m mod 3) - 1),
    // with m the current UTC month (1..12). See
    // docs/sample-benchmark.md test 4.
    const m = new Date().getUTCMonth() + 1; // getUTCMonth is 0-based
    const sleepS = 2.15 + ((m % 3) - 1);
    await sleep(Math.round(sleepS * 1000));
  });
});
